### PR TITLE
[Improvement] Invoke mkldnn and cudnn BatchNorm when axis != 1

### PR DESCRIPTION
**Pull request by wkcn**
## Description ##
Hi there.
In previous version, the MKLDNN and CUDNN implementation of BatchNorm  will be invoked when axis is 1. In other axis, the naive implementation will be applied.

In this PR, it support calling MKLDNN and CUDNN Batch Norm in arbitrary axis. 

If the shape of input `A` is `shape`, the input will be shaped into `(prod(shape[0:axis]), shape[axis], 1, prod(shape[axis+1:len(shape)]))`.

The PR contains the PR #18500, and it should be merged after #18500 .

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Feature1, tests, (and when applicable, API doc)
- [x] Feature2, tests, (and when applicable, API doc)

## Comments ##
The unittest `test_symbol.py` seems to be removed in the previous PR.
~I add a new param `mkldnn_off` into BatchNorm, since it is not numerical stable when the input is  extremely large. It will leads to the failure of `tests/python/unittest/test_symbol.py:test_load_000800`~
```
[2020-06-11T06:53:50.575Z] Items are not equal:
[2020-06-11T06:53:50.575Z] Error 1.072210 exceeds tolerance rtol=1.000000e-03, atol=1.000000e-03 (mismatch at least 5.500000%).
[2020-06-11T06:53:50.575Z] Location of maximum error: (0, 75), a=-12467010038075455310653024108544.00000000, b=-12480391637972769640987846770688.00000000
[2020-06-11T06:53:50.575Z]  ACTUAL: array([[ 2.9346620e+32,  1.9349947e+33, -1.0458480e+33, ...,
[2020-06-11T06:53:50.575Z]          2.2364499e+33,  1.0379849e+33,  5.1014372e+32]], dtype=float32)
[2020-06-11T06:53:50.575Z]  DESIRED: array([[ 2.9377518e+32,  1.9370348e+33, -1.0469506e+33, ...,
[2020-06-11T06:53:50.575Z]          2.2388062e+33,  1.0390785e+33,  5.1068152e+32]], dtype=float32)
```

### CODE REVIEW

**Comment by wkcn** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by wkcn** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by wkcn** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by wkcn** [comment]
@szha 
Thanks for your suggestion!
I agree. I will update this PR after replacing `mkldnn_off` and `cudnn_off` attributes with the environment variables.

**Comment by wkcn** [comment]
Hi @szha , could the PR be merged before replacing mkldnn_off and cudnn_off attributes with environment variables? I can remove the attribute mkldnn_off.

**Comment by szha** [comment]
👌🏻

**Comment by TaoLv** [comment]
@wkcn, do you have any performance number?

**Comment by wkcn** [comment]
> @wkcn, do you have any performance number?

Sorry that I do not have any performance table. I have only a laptop computer with i7-7500u 2C4T. Could you help me for the benchmark? Thank you!

**Comment by wkcn** [comment]
@mxnet-bot run ci [unix-cpu, centos-cpu]

**Comment by wkcn** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by sandeep-krishnamurthy** [comment]
@mseth10 @access2rohit - Please help review this PR.

File: src/operator/nn/cudnn/cudnn_batch_norm-inl.h [hunk 1]
```diff
@@ -262,15 +262,26 @@ class CuDNNBatchNormOp {
...
-      // when in_data.ndim() != 4
-      shape_[0] = in_data.shape_[0];
-      shape_[1] = in_data.ndim() > 1 ? in_data.shape_[1] : 1;
+      // reshape to (N, C, 1, D), C is the `param_.axis` dimension
+      shape_[0] = static_cast<dim_t>(in_data.shape_.ProdShape(0, param_.axis));
+      shape_[1] = in_data.shape_[param_.axis];
       shape_[2] = 1;
-      shape_[3] = in_data.shape_.ProdShape(2, in_data.ndim());
+      shape_[3] = static_cast<dim_t>(in_data.shape_.ProdShape(param_.axis + 1,
+            static_cast<int>(in_data.ndim())));
```
Comment (mseth10) [inline src/operator/nn/cudnn/cudnn_batch_norm-inl.h:284]: why do we need static_cast here? why cant we do it like line 276?

File: src/operator/nn/batch_norm.cc [hunk 2]
```diff
@@ -420,10 +420,14 @@ static bool BatchNormType(const nnvm::NodeAttrs& attrs,
 
 #if MXNET_USE_MKLDNN == 1
 static inline bool SupportMKLDNNBN(const NDArray &input, const BatchNormParam &param) {
-  mxnet::TShape shape = input.shape();
-  return SupportMKLDNN(input) && shape.ndim() == 4
```
Comment (mseth10) [inline src/operator/nn/batch_norm.cc:424]: we are removing the check for `ndim == 4` here, and another lighter check for `ndim == 1 || ndim ==2 || ndim ==4` present in `SupportMKLDNN`.
Does that mean `ndim` can be anything >0 ? What are the allowed values for `ndim`?

File: src/operator/nn/cudnn/cudnn_batch_norm-inl.h (hunk 1 above)
Comment (wkcn) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:284]: The return dtype of `ProdShape` is `size_t`, a unsigned int, but the dtype of shape_[i] is `dim_t`, namely `int64_t`.
I do not know whether the `static_cast` is neceseary to avoid the potential compiler's warning.

File: src/operator/nn/batch_norm.cc (hunk 2 above)
Comment (wkcn) [reply src/operator/nn/batch_norm.cc:424]: Yes, `ndim` coulld be anything > 0.

If the shape of input A is `shape`, the input will be shaped into `(prod(shape[0:axis]), shape[axis], 1, prod(shape[axis+1:len(shape)]) )`.

ndim|shape|axis=0|axis=1|axis=2|axis=3|axis=4
---|---|--|--|--|--|---
1|(N,)|(1, N, 1, 1)|x|x|x|x
2|(N,C)|(1, N, 1, C)|(N, C, 1, 1)|x|x|x
3|(N,C,H)|(1, N, 1, C*H)|(N, C, 1, H)|(N*C, H, 1, 1)|x|x
4|(N,C,H,W)|(1, N, 1, C*H*W)|(N, C, 1, H*W)|(N*C, H, 1, W)|(N*C*H, W, 1, 1)|x
5|(N,D,C,H,W)|(1, N, 1, D*C*H*W)|(N, D, 1, C*H*W)|(N*D,C, 1, H*W)|(N*D*C, H, 1, W)|(N*D*C*H, W, 1, 1)

Comment (mseth10) [reply src/operator/nn/batch_norm.cc:424]: Thanks for the explanation! Does the table continue further for ndims > 5? Or should we place a check for that?

Comment (wkcn) [reply src/operator/nn/batch_norm.cc:424]: Yes, it supports ndim > 5 too.

File: src/operator/nn/cudnn/cudnn_batch_norm-inl.h (hunk 1 above)
Comment (mseth10) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:284]: Can we add static_cast to line 276 as well?

Comment (mseth10) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:284]: in_data.ndims() shouldn’t need a static cast though, right?

Comment (wkcn) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:284]: Yes, the dtype of `.ndim()` is `int32_t`.

variable|dtype
---|---
ProdShape | size_t
shape_[i] | dim_t, int64_t
ndim() | int32_t

Comment (wkcn) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:284]: Hi @mseth10 , do you have any suggestion about whether to use `static_cast` ?

Comment (mseth10) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:284]: Imo you can ignore the compiler warnings (if any) for int32_t to int,
you can keep static_cast for size_t to dim_t

Comment (wkcn) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h]: I see. I have updated it : )

**Comment by stu1130** [comment]
@mxnet-bot run ci [macosx-x86_64]

**Comment by stu1130** [comment]
@TaoLv could you help merge it if the PR looks good to you?

**Comment by sandeep-krishnamurthy** [comment]
@szha @eric-haibin-lin - Can you please help review/merge with this PR? Thank you.
